### ISSUE TITLE
deleting key before executing transaction

### ISSUE THREAD

**Issue by klem4** [jamesls/fakeredis#124]
Hi! I think i found a bug. In my example, i want to make a transaction with combination of GET and DELETE methods, but because pipe.delete() runs redis.delete command immediately, when the execute are called, key already deleted from DB!

test script:

``` python
# coding: utf-8

from redis import Redis
from fakeredis import FakeRedis

KEY = 'x'
VALUE = '1'

r = FakeRedis() #Redis()
r.set(KEY, VALUE)
assert r.get(KEY) == VALUE


def get_check_delete(pipe):
    real_value = pipe.get(KEY)
    if not real_value == VALUE:
        raise ValueError(KEY)
    pipe.delete(KEY)


print r.transaction(get_check_delete, KEY)
```

problem code i think:
fakeredis.py:1144 and fakeredis.py:1209

``` python
def transaction(self, func, *keys):
        with self.pipeline() as p:
            for _ in range(5):
                try:
                    p.watch(*keys)
                    func(p) <-------------- key deleted here!
                    return p.execute() <--------------- deleted key unavailable in execute!
...
```

my example code based on redis's lock mechanism:
redis/lock.py:108

``` python
def do_release(self):
        name = self.name
        token = self.token

        def execute_release(pipe):
            lock_value = pipe.get(name)
            if lock_value != token:
                raise LockError("Cannot release a lock that's no longer owned")
            pipe.delete(name)

        self.redis.transaction(execute_release, name)
```

if you agree there it is bug, i can try to create pull requests and tests to fix it.

**Comment by eswald** [comment]
There's a nasty interaction here between `FakePipeline.watching` and `FakePipeline.is_immediate` when the transaction modifies or deletes a watched key.  The pipeline's `execute` method notices that the value was changed (because `func(p)` changed it), and fires a WatchError.  That error gets swallowed by the loop inside `transaction`, which then launches the transaction a second time with the value already changed.

Two issues, then: 1) WatchError should only fire if the key is changed _outside_ of the transaction, and 2) the transaction should only be retried if it all side effects from a previous run can be rolled back.

Meanwhile, there's a workaround for this particular case: Running `pipe.multi()` before deleting or modifying a watched variable defers the changes until after the variable watch check.

**Comment by bmerry** [comment]
This is fixed in fakeredis 0.14.0.

`pipe.multi` isn't just a workaround, it's necessary to make the process transactional. The redis.lock.Lock implementation actually has a race condition here (https://github.com/andymccurdy/redis-py/pull/1031).
